homeassistant/components/mazda/binary_sensor.py

```diff
@@ -0,0 +1,165 @@
+"""Platform for Mazda binary sensor integration."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Any
+
+from homeassistant.components.binary_sensor import (
+    BinarySensorDeviceClass,
+    BinarySensorEntity,
+    BinarySensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import MazdaEntity
+from .const import DATA_CLIENT, DATA_COORDINATOR, DOMAIN
+
+
+@dataclass
+class MazdaBinarySensorRequiredKeysMixin:
+    """Mixin for required keys."""
+
+    # Suffix to be appended to the vehicle name to obtain the binary sensor name
+    name_suffix: str
+
+    # Function to determine the value for this binary sensor, given the coordinator data
+    value: Callable[[dict[str, Any]], bool]
+
+
+@dataclass
+class MazdaBinarySensorEntityDescription(
+    BinarySensorEntityDescription, MazdaBinarySensorRequiredKeysMixin
+):
+    """Describes a Mazda binary sensor entity."""
+
+    # Function to determine whether the vehicle supports this binary sensor, given the coordinator data
+    is_supported: Callable[[dict[str, Any]], bool] = lambda data: True
+
+    # Function to return extra state attributes, given the coordinator data
+    state_attributes: Callable[[dict[str, Any]], dict[str, Any]] | None = None
+
+
+def _plugged_in_supported(data):
+    """Determine if 'plugged in' binary sensor is supported."""
+    return (
+        data["isElectric"] and data["evStatus"]["chargeInfo"]["pluggedIn"] is not None
+    )
+
+
+def _charging_supported(data):
+    """Determine if 'charging' binary sensor is supported."""
+    return data["isElectric"] and data["evStatus"]["chargeInfo"]["charging"] is not None
+
+
+def _doors_open(data):
+    """Determine if at least one vehicle door is open."""
+    door_status = data["status"]["doors"]
+    return (
+        door_status["driverDoorOpen"]
+        or door_status["passengerDoorOpen"]
+        or door_status["rearLeftDoorOpen"]
+        or door_status["rearRightDoorOpen"]
+    )
+
+
+def _doors_state_attributes(data):
+    """Get state attributes for the vehicle doors."""
+    door_status = data["status"]["doors"]
+
+    return {
+        "driver_door_open": door_status["driverDoorOpen"],
+        "passenger_door_open": door_status["passengerDoorOpen"],
+        "rear_left_door_open": door_status["rearLeftDoorOpen"],
+        "rear_right_door_open": door_status["rearRightDoorOpen"],
+    }
```

Comment:
I can't quite decide if it is ok to have these as attributes? or if they should they be four separate sensors?

Reply:
Hmm, why wouldn't it be ok to have them as attributes? According to the developer docs, extra state attributes should be "information that further explains the state", which is what I think I've done here. The entity state says that the car has a door open, and the state attributes further explain that by specifying which door is open.

Reply:
I'll let @frenck or another review confirm, but we are generally moving away from attributes. If you look at the commit history you will find quite a few commits to "migrate attributes to individual sensors".
It's just that in this particular case I can't quite decide...

Reply:
Fair enough. The main reason I made it one sensor is because I thought it would make it easier for users to create automations. If they were four separate sensors then users would have to create an "or" condition every time. I was also trying not to clutter the UI with too many entities. But if another reviewer confirms that attributes should not be used here, I can change it to individual sensors.

Reply:
Giving my two cents, as i have made a custom integration for Toyota. I use individual sensors for each door. I can see where @bdr99 comes from, and I also thought long about what was the best option, as the user probably does not care _which_ door is open, but rather _if_ a door is open. My integration creates a lot of entities for just a single car, 25+. On the other hand, I personally prefer individual sensors over attributes, as attributes are harder to work with in the UI and in automations. (Though it have become easier over the last year).